alf/optimizers/optimizers.py

```diff
@@ -92,8 +181,103 @@ def step(self, closure=None):
+                    [p.view(batch_size, -1) for p in params],
+                    dim=-1)  # [N, D], D=dim(params)
+                if self._parvi == 'svgd':
+                    # [N, N], [N, N, D]
+                    kappa, kappa_grad = rbf_func(params_tensor)
+                    grads_tensor = torch.cat(
+                        [p.grad.view(batch_size, -1) for p in params],
+                        dim=-1).detach()  # [N, D]
+                    kernel_logp = torch.matmul(kappa,
+                                               grads_tensor) / batch_size
+                    svgd_grad = torch.split(
+                        kernel_logp -
+                        self._repulsive_weight * kappa_grad.mean(0),
+                        [int(p.nelement() / batch_size) for p in params],
+                        dim=-1)
+                    for i in range(len(params)):
+                        grad = params[i].grad.view(batch_size, -1).zero_()
+                        grad.add_(svgd_grad[i])
+                else:
+                    logq_grad = score_func(params_tensor)  # [N, D]
+                    gfsf_grad = torch.split(
+                        logq_grad,
+                        [int(p.nelement() / batch_size) for p in params],
+                        dim=-1)
+                    for i in range(len(params)):
+                        grad = params[i].grad.view(batch_size, -1)
+                        grad.add_(self._repulsive_weight * gfsf_grad[i])
+
+    @common.add_method(NewCls)
+    def add_param_group(self, param_group):
+        """This function first splits the input param_group into multiple
+        param_groups according to their ``ensemble_group`` attributes, then 
+        calls the parent's ``add_param_group()`` function to add each of
+        them to the optimizer.
+        """
+        assert isinstance(param_group, dict), "param_group must be a dict"
+
+        params = param_group["params"]
+        if isinstance(params, torch.Tensor):
+            param_group['params'] = [params]
+        elif isinstance(params, set):
+            raise TypeError('Please use a list instead.')
+        else:
+            param_group['params'] = list(params)
+
+        len_params = len(param_group['params'])
+        std_param_group = []
+        ensemble_param_groups = [[] for i in range(len_params)]
+        group_batch_sizes = torch.zeros(len_params)
+        for param in param_group['params']:
+            if not isinstance(param, torch.Tensor):
+                raise TypeError("optimizer can only optimize Tensors, "
+                                "but one of the params is " +
+                                torch.typename(param))
+            if hasattr(param, 'ensemble_group'):
+                assert isinstance(
+                    param.ensemble_group,
+                    int), ("ensemble_group attribute mis-specified.")
+                ensemble_group_id = param.ensemble_group % len_params
```

Comment:
why do you need '%'?

Reply:
In case the attribute "ensemble_group" is specified as some int > len_params? Anyway, I think the programmer should be aware of this. Just removed.